module/zfs/sa.c

```diff
@@ -1496,6 +1500,185 @@ sa_lookup_uio(sa_handle_t *hdl, sa_attr_type_t attr, uio_t *uio)
 	mutex_exit(&hdl->sa_lock);
 	return (error);
 }
+
+static int
+sa_lookup_locked(sa_handle_t *hdl, sa_attr_type_t attr, void *buf,
+    uint32_t buflen)
+{
+	int error;
+	sa_bulk_attr_t bulk;
+
+	VERIFY3U(buflen, <=, SA_ATTR_MAX_LEN);
+
+	bulk.sa_attr = attr;
+	bulk.sa_data = buf;
+	bulk.sa_length = buflen;
+	bulk.sa_data_func = NULL;
+
+	ASSERT(hdl);
+	error = sa_lookup_impl(hdl, &bulk, 1);
+	return (error);
+}
+
+int
+sa_add_projid(sa_handle_t *hdl, dmu_tx_t *tx, uint64_t projid)
```

Comment:
It seems more appropriate to place this function in `module/zfs/zfs_sa.c` and rename it `zfs_sa_set_projid()`.  That would allow it to be restructured along the lines of `zfs_sa_set_scanstamp()` and rely on the `sa_update()->zfs_sa_upgrade()` path to perform the upgrade when needed.  Or did you already try this approach and run in to problems?